frontends/p4/simplifyDefUse.cpp

```diff
@@ -271,8 +451,63 @@ class FindUninitialized : public Inspector {
     bool preorder(const IR::P4Parser* parser) override {
         LOG3("FU Visiting parser " << parser->name << "[" << parser->id << "]");
         currentPoint = ProgramPoint(parser);
+        headerDefs->clear();
+        initHeaderParams(parser->getApplyMethodType()->parameters);
         visitVirtualMethods(parser->parserLocals);
-        visit(parser->states, "states");
+
+        auto startState = parser->getDeclByName(IR::ParserState::start)->to<IR::ParserState>();
+        auto acceptState = parser->getDeclByName(IR::ParserState::accept)->to<IR::ParserState>();
+
+        ParserCallGraph transitions("transitions");
+        ComputeParserCG pcg(refMap, &transitions);
+
+        (void)parser->apply(pcg);
+        ordered_set<const IR::ParserState*> toRun;  // worklist
+        ordered_map<const IR::ParserState*, HeaderDefinitions*> inputHeaderDefs;
+
+        toRun.emplace(startState);
+        inputHeaderDefs.emplace(startState, headerDefs);
+
+        // We do not report warnings until we have all definitions for every parser state
+        reportInvalidHeaders = false;
+
+        while (!toRun.empty()) {
+            auto state = *toRun.begin();
+            toRun.erase(state);
+            LOG3("Traversing " << dbp(state));
+
+            // We need a new visitor to visit the state,
+            // but we use the same data structures
+            headerDefs = inputHeaderDefs[state]->clone();
+            FindUninitialized fu(this, currentPoint);
+            (void)state->apply(fu);
+
+            auto next = transitions.getCallees(state);
+            for (auto n : *next) {
+                if (inputHeaderDefs.find(n) == inputHeaderDefs.end()) {
+                    inputHeaderDefs[n] = headerDefs->clone();
+                    toRun.emplace(n);
+                } else {
+                    auto joined = inputHeaderDefs[n]->intersect(headerDefs);
```

Comment:
usually "join" means "union", so perhaps you want to rename this variable to something like 'common'.